File: typespec/hub/comments.ts
export class AddPostCommentRequest {
  post_id: string = "";
  content: string = "";

  IsValid(): boolean {
    return (
      this.post_id.length > 0 &&
      this.content.length > 0 &&
      this.content.length <= 4096
    );
  }
}

export interface AddPostCommentResponse {
  post_id: string;
  comment_id: string;
}

export class GetPostCommentsRequest {
  post_id: string = "";
  pagination_key?: string;
  limit?: number;
  IsValid(): boolean {
    return (
      this.post_id.length > 0 &&
      (this.limit === undefined || (this.limit >= 0 && this.limit <= 40))
    );
  }
}

export interface PostComment {
  id: string;
  content: string;
  author_name: string;
  author_handle: string;
  created_at: Date;
}

export class DisablePostCommentsRequest {
  post_id: string = "";
  delete_existing_comments: boolean = false;

  IsValid(): boolean {
    return this.post_id.length > 0;
  }
}

export class EnablePostCommentsRequest {
  post_id: string = "";

  IsValid(): boolean {
    return this.post_id.length > 0;
  }
}

export class DeletePostCommentRequest {
  post_id: string = "";
  comment_id: string = "";

  IsValid(): boolean {
    return this.post_id.length > 0 && this.comment_id.length > 0;
  }
}

export class DeleteMyCommentRequest {
  post_id: string = "";
  comment_id: string = "";

  IsValid(): boolean {
    return this.post_id.length > 0 && this.comment_id.length > 0;
  }
}

File: typespec/hub/comments.go
package hub

import (
	"time"

	"github.com/vetchium/vetchium/typespec/common"
)

type AddPostCommentRequest struct {
	PostID  string `json:"post_id" validate:"required"`
	Content string `json:"content" validate:"required,min=1,max=4096"`
}

type AddPostCommentResponse struct {
	PostID    string `json:"post_id"`
	CommentID string `json:"comment_id"`
}

type GetPostCommentsRequest struct {
	PostID        string `json:"post_id"        validate:"required"`
	PaginationKey string `json:"pagination_key"`
	Limit         int    `json:"limit"          validate:"min=0,max=40"`
}

type PostComment struct {
	ID           string        `json:"id"`
	Content      string        `json:"content"`
	AuthorName   string        `json:"author_name"`
	AuthorHandle common.Handle `json:"author_handle"`
	CreatedAt    time.Time     `json:"created_at"`
}

type DisablePostCommentsRequest struct {
	PostID                 string `json:"post_id"                  validate:"required"`
	DeleteExistingComments bool   `json:"delete_existing_comments"`
}

type EnablePostCommentsRequest struct {
	PostID string `json:"post_id" validate:"required"`
}

type DeletePostCommentRequest struct {
	PostID    string `json:"post_id"    validate:"required"`
	CommentID string `json:"comment_id" validate:"required"`
}

type DeleteMyCommentRequest struct {
	PostID    string `json:"post_id"    validate:"required"`
	CommentID string `json:"comment_id" validate:"required"`
}
